domain/upgrade/state/state.go

```diff
@@ -0,0 +1,304 @@
+	if err == sqlair.ErrNoRows {
+		return info{}, nil, noUpgradeErr
+	}
+	if err != nil {
+		return info{}, nil, errors.Trace(err)
+	}
+
+	var nodeInfos []infoControllerNode
+	err = tx.Query(ctx, s2, sqlair.M{"info_uuid": upgradeInfo.UUID}).GetAll(&nodeInfos)
+	if err != nil && err != sqlair.ErrNoRows {
+		return info{}, nil, errors.Trace(err)
+	}
+	return upgradeInfo, nodeInfos, nil
+}
+
+// CreateUpgrade creates an upgrade to and from specified versions.
+// If an upgrade is already running/pending, return an error
+func (st *State) CreateUpgrade(ctx context.Context, previousVersion, targetVersion version.Number) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		info, _, err := st.getCurrentUpgrade(ctx, tx)
+		if err != nil && err != noUpgradeErr {
+			return errors.Trace(err)
+		}
+		if err == nil {
+			return errors.AlreadyExistsf("upgrade %s -> %s", info.PreviousVersion, info.TargetVersion)
+		}
+
+		q := `
+INSERT INTO upgrade_info (uuid, previous_version, target_version, created_at) VALUES
+($M.uuid, $M.previousVersion, $M.targetVersion, DATETIME('now'))`
+		s, err := sqlair.Prepare(q, sqlair.M{})
+		if err != nil {
+			return errors.Annotatef(err, "preparing %q", q)
+		}
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{
+			"uuid":            utils.MustNewUUID().String(),
+			"previousVersion": previousVersion.String(),
+			"targetVersion":   targetVersion.String(),
+		}).Run())
+	}))
+}
+
+// SetControllerReady marks the supplied controllerID as being ready
+// to start it's upgrade. All provisioned controllers need to be ready
+// before an upgrade can start
+// a controller is ready for a particualry upgrade if it's ID is present
+func (st *State) SetControllerReady(ctx context.Context, controllerID string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		info, nodeInfos, err := st.getCurrentUpgrade(ctx, tx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, nodeinfo := range nodeInfos {
+			if nodeinfo.ControllerNodeID == controllerID {
+				return nil
+			}
+		}
+		q := `
+INSERT INTO upgrade_info_controller_node (uuid, controller_node_id, upgrade_info_uuid) VALUES
+($M.uuid, $M.controllerID, $M.infoUUID)`
+		s, err := sqlair.Prepare(q, sqlair.M{})
+		if err != nil {
+			return errors.Annotatef(err, "preparing %q", q)
+		}
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{
+			"uuid":         utils.MustNewUUID().String(),
+			"controllerID": controllerID,
+			"infoUUID":     info.UUID,
+		}).Run())
+	}))
+}
+
+// AllProvisionedControllersReady returns true if and only if all controllers
+// that have been started by the provisioner have called EnsureUpgradeinfo with
+// matching versions.
+func (st *State) AllProvisionedControllersReady(ctx context.Context) (bool, error) {
```

Comment:
This should accept the current upgrade UUID, and do an outer join of `upgrade_info_controller_node` with `controller_node` to make this determination in a single query.